Classes/EventListener/CheckProductAvailability.php

```diff
@@ -10,86 +10,107 @@
 
-        $querySettings = $this->productRepository->createQuery()->getQuerySettings();
-        $querySettings->setRespectStoragePage(false);
-        $this->productRepository->setDefaultQuerySettings($querySettings);
+        if ($cartProduct->isHandleStockInVariants()) {
+            foreach ($cartProduct->getBeVariants() as $cartProductBeVariant) {
+                $this->checkStockForBeVariant($cartProductBeVariant, $cartProduct);
+            }
+        } else {
+            $this->checkStockForProduct($cartProduct);
+        }
+    }
 
-        $product = $this->productRepository->findByIdentifier($cartProduct->getProductId());
+    private function checkStockForProduct(CartProduct $cartProduct): void
+    {
+        $cart = $this->event->getCart();
+        $mode = $this->event->getMode();
+
+        $quantity = $this->event->getQuantity();
+        // $quantity is an array in the event that the number of products with variants is updated in the shopping cart.
+        // In this case, the number from the update request must be added up, as stock management for all variants
+        // together takes place in the product.
+        // $quantity is not an array if a product or a product variant is added via the product page. In this case, the
+        // value for the quantity must be taken directly.
+        if (is_array($quantity)) {
+            $quantityInCart = array_sum($quantity);
+        } else {
+            $quantityInCart = (int)$quantity;
+        }
 
-        if (
-            !$product instanceof Product ||
-            !$product->isHandleStock()
-        ) {
-            return;
+        if (($mode === 'add') && $cart->getProductById($cartProduct->getId())) {
+            $quantityInCart += $cart->getProductById($cartProduct->getId())->getQuantity();
         }
 
-        if (!$product->isHandleStockInVariants()) {
-            if (is_array($quantity)) {
-                $compareQuantity = array_sum($quantity);
-            } else {
-                $compareQuantity = $quantity;
-            }
-            if (($mode === 'add') && $cart->getProductById($cartProduct->getId())) {
-                $compareQuantity += $cart->getProductById($cartProduct->getId())->getQuantity();
-            }
+        $quantityInStock = $this->productRepository->getStock($cartProduct->getProductId());
 
-            if ($compareQuantity > $product->getStock()) {
-                $this->falseAvailability($event);
-            }
+        if ($quantityInStock < $quantityInCart) {
+            $this->falseAvailability();
+        }
+    }
 
-            return;
+    public function checkStockForBeVariant(CartProductBeVariant $cartProductBeVariant, CartProduct $cartProduct): void
+    {
+        $cart = $this->event->getCart();
+        $mode = $this->event->getMode();
+
+        $quantity = $this->event->getQuantity();
+        // $quantity is an array in the event that the number of products with variants is updated in the shopping cart.
```

Comment:
Proposal: 
```php
// $quantity is an array in case that the number of a products with variants
// is updated within the shopping cart.
```

Reply:
I've changed the wording of the first sentence.